### ISSUE TITLE
vat termination

### ISSUE THREAD

**Issue by warner** [Agoric/agoric-sdk#1466]
**What is the Problem Being Solved?**

We want to reduce the amount of memory and disk used by a long-running swingset. One thing that will help is to separate objects into vats which can be terminated when their task is done. The simplest example is a basic exchange contract: once the atomic swap is complete, the contract is no longer needed. We run the contract and any auxilliary objects in a single vat, and terminate the vat afterwards. At the very least the VatWorker can be released and GCed (saving RAM), and the vat transcript can be discarded (saving disk). In addition, any references leading out of the dead vat are clearly no longer needed, giving us an explicit GC `decref` that can help remove c-list and kernel table entries.

(eventually we hope to have JS-level weakrefs to give us more decref signals to work with, but making them deterministic is pretty tricky, so deleting a whole vat at a time is a practical mitigation)

Vats can be killed explicitly, when the creator invokes the admin facet's `terminate()` method, or through failure (metering underflow, or an internal vat error which causes it to make an illegal syscall, like resolving a Promise twice, or referencing a missing c-list import). In either case, there are three main behavioral features we want:

* the dead vat loses agency: despite everything the vat does before its death, and everything other vats do afterwards, its code will never be invoked again
  * all messages sent to objects in the vat should be rejected with some sort of "DeadVat" error object
* the creator's `done` promise is settled: resolved (perhaps with some status object) if the vat was killed deliberately via `terminate()`, or rejected if the cause of death was a metering or internal fault
* all Promises decided by the now-dead vat should be rejected with the same DeadVat error as inbound messages

Our current metering implementation only does the first half. We have no `terminate()` call, but vats can be killed by metering faults. Once a vat underflows its meter, the meter stops being refilled. Messages will continue to be delivered to the vat, but they will all throw an underflow error immediately. This effectively removes the vat's agency, but doesn't allow us to clean anything up. In addition, the late vat's Promises are not currently rejected.

In addition to that basic functionalty, we want actual performance improvements. We anticipate several stages:

* 1: we don't actually clean anything up, we just make sure messages and promises are rejected.
* 2: delete most of the dead vat's kernel state, and release the VatManager
* 3: try to delete kernel object table entries for the ones that lived in the dead vat. We can send obituaries to other vats, which can maybe let them delete the strong-map (object-id to Presence) direction of their liveslots tables, and if the vat had already dropped the Presence (releasing the weak-map direction), then actual GC can take place
* 4: do the same thing but over the comms connections to remote machines

Promises might be easier to clean up because 1: we already have the beginnings of code to do that, and 2: Promises don't have as much identity as objects. We have to make some semantic decisions here. In the E world, object references retain identity even if they become Broken. I'm guessing we want to retain that property here, and that might require us to retain the kernel object table entry for as long as any vat's c-list references it, which might mean we can't really clean this up until we have JS-level (`WeakRef`) signals that the vat has become incapable of referencing the now-dead object.

**Description of the Design**

For our first stage, we'll add a `vNN.terminated` flag to the durable kernelDB, and an `ephemeral.vats.get(NN).terminated` flag to the in-RAM record. When the vat is terminated (either through metering or explicitly), we set both flags and notify the vat-admin watcher (which will queue a resolution for the `done` promise). We'll change `kernel.deliverToVat` to check for the `.terminated` flag before talking to the vatManager. If set, and if the message has a result promise, it needs to synthesize a rejection. We change `notify` to ignore notifications for dead vats.

(We'll retain that durable `vNN.terminated` flag in the DB until we've finished all the other stages and can be sure that the `vNN` identifier is not referenced anywhere in the kernel tables. At that point we can forget entirely about the old vat, but until then we want to be able to distinguish between a bug causing a reference to a non-existent vat ID, and a stale reference pointing at a now-dead vat ID.)

Then, we must somehow enumerate all the promises that are controlled by the dead vat (`p.state === 'unresolved' && p.decider === deadVatID`). This is probably expensive right now, because our (simple KV-store) DB isn't designed to support this sort of query. One possible direction is to walk the entire kernel promise table (all `kpid` integers from 0 to the current counter, query for `pNN.state` for each). A slightly cheaper approach would be to walk the dead vat's clist and look for all the promises it ever had access to. If/when we move to a different DB system, maybe SQLite, we'd want an index on the decider row to make this fast.

For each such promise, we pretend the vat just did a `syscall.reject`. This will change its state (from `unresolved` to `rejected`) and queue a `notify` for all its subscribers. It will also delete the `vpid` from the vat's clist, but we could elect to skip that step because the entire clist is about to be deleted (or retain it for the sake of code reuse). The promise might be deleted from the kernel table if our GC rules allow it, either immediately or shortly afterwards (once other references go away).

cc @FUDCo 

**Test Plan**

We need to list all the places a vat ID might be stored, and test what happens with each of them after the vat has been terminated. I can think of: c-lists, kernel object table, the "decider" of an unresolved kernel promise table entry, and a run-queue "notify" task (which cites a specific vat ID to be notified about some promise ID's new state).

**Comment by warner** [comment]
Some notes from our call:

* the `ephemeral.vats` record can remain hardened, but we'll replace it with a different copy when the vat is dead. It will be born with a `dead: false` property, and the replacement will have `dead: true` (and no `.manager` or `.translators`

```js
  async function removeVatManager(vatID) {
    const old = ephemeral.vats.get(vatID);
    // once record.dead is set, no further messages will make it to the
    // manager, so it's safe to remove now
    ephemeral.vats.set(vatID, harden({dead: true}));
    // shutdown might take an unknown length of time (killing a child thread/process)
    old.manager.shutdown().catch(log at a high level, but not fatal)
  }
```

* merge `kernelSyscall.js reject(vatID, kpid, data)` and `kernel deliverToError(kpid, errorData, vatToNotNotify)`

* `deliverToVat` should check `ephemeral.vats.get(vatID).dead` and call `deliverToError` if so

```js
  async function deliverToVat(vatID, target, msg) {
    insistMessage(msg);
    const vat = ephemeral.vats.get(vatID);
    if (vat.dead) {
      const s = `vat is dead`; // note: don't include vatID, for determinism/isolation
      await deliverToError(msg.result, makeError(s));
    }
    assert(vat, details`unknown vatID ${vatID}`);
   ...
```

* when we load vats from the kernelDB at restart, look for the durable/persistent `vNN.dead` flag and skip that vat entirely

**Comment by FUDCo** [comment]
Is this done now?

**Comment by warner** [comment]
Yeah,this is pretty much done. The parts about deleting object-table entries will happen more naturally as part of the GC effort, so I don't think we need new tickets to cover them.
